### PR TITLE
refactor: Assert

### PR DESCRIPTION
**Pull request by shubham-cmyk**

### CODE REVIEW

File: pkg/runner/operations/assert/assert.go [hunk 1]
```diff
@@ -47,41 +48,18 @@ func (o *operation) Exec(ctx context.Context) (_err error) {
...
-		}()
-		if candidates, err := internal.Read(ctx, &o.expected, o.client); err != nil {
-			if kerrors.IsNotFound(err) {
-				errs = append(errs, errors.New("actual resource not found"))
-				return false, nil
-			}
+	err := wait.PollUntilContextCancel(ctx, internal.PollInterval, false, func(ctx context.Context) (bool, error) {
+		candidates, errs, err := o.fetchAndValidateCandidates(ctx)
+		if err != nil || len(errs) != 0 {
+			lastErrs = errs
```
[the remark this answers is not in the document]
Comment (shubham-cmyk) [reply pkg/runner/operations/assert/assert.go:59]: let me check it passed the test case

[the remark this answers is not in the document]
Comment (eddycharly) [reply pkg/runner/operations/assert/assert.go:59]: a test may be missing

[the remark this answers is not in the document]
Comment (shubham-cmyk) [reply pkg/runner/operations/assert/assert.go:59]: let me revisit the logic. When it could fail

[the remark this answers is not in the document]
Comment (eddycharly) [reply pkg/runner/operations/assert/assert.go:59]: when the timeout is reached, we don't want to return the timeout error, we want to return the errors that made the assertion fail.

[the remark this answers is not in the document]
Comment (eddycharly) [reply pkg/runner/operations/assert/assert.go:59]: yes, it will

File: pkg/runner/operations/assert/assert.go [hunk 2]
```diff
@@ -47,41 +48,20 @@ func (o *operation) Exec(ctx context.Context) (_err error) {
...
 	var lastErrs []error
-	err := wait.PollUntilContextCancel(ctx, internal.PollInterval, false, func(ctx context.Context) (_ bool, err error) {
-		var errs []error
-		defer func() {
-			// record last errors only if there was no real error
-			if err == nil {
+	err := wait.PollUntilContextCancel(ctx, internal.PollInterval, false, func(ctx context.Context) (bool, error) {
+		candidates, errs, err := o.fetchAndValidateCandidates(ctx)
+		if err != nil || len(errs) != 0 {
+			if !kerrors.IsTimeout(err) {
```
Comment (eddycharly) [inline pkg/runner/operations/assert/assert.go:59]: is this error returned when the context is cancelled ?

Comment (shubham-cmyk) [reply pkg/runner/operations/assert/assert.go]: fixed

**Comment by eddycharly** [comment]
@shubham-cmyk i'm not convinced by the approach, maybe we should let handlers decide if the want to override the last errors or not 🤔

**Comment by shubham-cmyk** [comment]
okay

**Comment by eddycharly** [comment]
Closing, needs significant work

**Pull request closed without merging**